src/app/contexts/Ledger/use-ledger.ts

```diff
@@ -0,0 +1,182 @@
+
+const formatDerivationPath = (coinType: number, index: number) => `44'/${coinType}'/${index}'/0/0`;
+
+export type LedgerData = {
+	address: string;
+	balance: BigNumber;
+	index: number;
+	isNew?: boolean;
+};
+
+export const useLedger = (transport: typeof Transport) => {
+	const { env, persist } = useEnvironmentContext();
+	const [state, dispatch] = useReducer(ledgerStateReducer, defaultLedgerState);
+	const abortRetryRef = useRef<boolean>(false);
+
+	const { isBusy, isConnected, error } = state;
+
+	const listenDevice = useCallback(
+		() =>
+			transport.listen({
+				// @ts-ignore
+				next: ({ type, descriptor, deviceModel }) => {
+					if (type === "add") {
+						dispatch({ type: "add", path: descriptor, id: deviceModel?.id || "nanoS" });
+						return;
+					}
+					dispatch({ type: "remove" });
+				},
+				error: (e) => dispatch({ type: "failed", message: e.message }),
+				complete: () => void 0,
+			}),
+		[transport],
+	);
+
+	const importLedgerWallets = useCallback(
+		async (wallets: LedgerData[], coin: Coins.Coin, profile: Profile) => {
+			for (const { address, index } of wallets) {
+				const wallet = await profile
+					.wallets()
+					.importByAddress(address, coin.network().coin(), coin.network().id());
+				wallet.data().set(WalletFlag.LedgerIndex, index);
+			}
+			await persist();
+		},
+		[persist],
+	);
+
+	const connect = useCallback(
+		async (
+			coin: string,
+			network: string,
+			retryOptions: retry.Options = { retries: 50, randomize: false, factor: 1 },
+		) => {
+			dispatch({ type: "waiting" });
+			abortRetryRef.current = false;
+
+			const instance = await env.coin(coin, network);
+
+			try {
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				const connectFn: retry.RetryFunction<void> = async (bail, attempts) => {
+					if (abortRetryRef.current && attempts > 1) {
+						bail(new Error("User aborted"));
+					}
+
+					await instance.ledger().connect(transport);
+					// Ensure that the app is accessible
+					await instance.ledger().getPublicKey(formatDerivationPath(slip44, 0));
+				};
+
+				await retry(connectFn, retryOptions);
+				dispatch({ type: "connected" });
+			} catch (connectError) {
+				instance.ledger().disconnect();
+				dispatch({ type: "failed", message: connectError.message });
+				throw connectError;
+			}
+		},
+		[transport, env],
+	);
+
+	const disconnect = useCallback(async (coin: Coins.Coin) => {
+		await coin.ledger().disconnect();
+		dispatch({ type: "disconnected" });
+	}, []);
+
+	const scanWallets = useCallback(
+		async (coin: string, network: string, profile: Profile, onChange?: (wallet: LedgerData) => void) => {
+			const wallets: LedgerData[] = [];
+
+			try {
+				await connect(coin, network);
+
+				const instance = await env.coin(coin, network);
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				dispatch({ type: "busy" });
+
+				let hasMore = true;
+				let cursor = 0;
+
+				while (hasMore) {
```

Comment:
Since this `while` loop performs an API call it would be better to push these tasks into an array of promises and then use `Promise.allSettled` to resolve them faster instead of running 1 request at a time.

Reply:
The scan stops until it finds an empty wallet, in this case it will be searched until the first one with `status: rejected` & `reason:404`?

Reply:
This issue goes back to https://github.com/ArkEcosystem/desktop-wallet/pull/2926#discussion_r503025589. We shouldn't be aborting this process just because a few wallets threw a 404 or had a failing HTTP request for whatever reason. All the wallets that fail to be imported should simply be gathered and shown to the user at the end with a reason for why it failed and then let them retry.

Reply:
Ok, how many wallets must be fetched? Doesn't look good to show 50 empty wallets for the user to choose if they really want to import it